File: backend/app/src/main/kotlin/net/animeta/backend/metadata/WorkMetadata.kt
package net.animeta.backend.metadata

import com.fasterxml.jackson.annotation.JsonFormat
import net.animeta.backend.model.Period
import java.time.LocalDateTime

data class WorkMetadata(
    val version: Int = 2,
    val title: String? = null,
    val periods: List<Period>? = null,
    val studios: List<String>? = null,
    val source: SourceType? = null,
    val website: String? = null,
    val namuRef: String? = null,
    val annId: String? = null,
    val durationMinutes: Int? = null,
    val schedules: Map<String, Schedule>? = null
) {
    enum class SourceType {
        MANGA,
        ORIGINAL,
        LIGHT_NOVEL,
        GAME,
        FOUR_KOMA,
        VISUAL_NOVEL,
        NOVEL,
    }

    data class Schedule(
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        val date: LocalDateTime?,
        val datePrecision: DatePrecision?,
        val broadcasts: List<String>?
    ) {
        enum class DatePrecision {
            YEAR_MONTH,
            DATE,
            DATE_TIME,
        }
    }
}

File: web/frontend/js/admin/WorkMetadata.ts
import { SourceType } from "../../../shared/client";

export interface WorkMetadata {
  periods?: string[] | null;
  website?: string | null;
  namuRef?: string | null;
  jaWikipediaRef?: string | null;
  annId?: string | null;
  source?: SourceType | null;
  studios?: string[] | null;
  durationMinutes?: number | null;
  schedules?: {[country: string]: Schedule} | null;
  _comment?: string;
}

export type DatePrecision = 'YEAR_MONTH' | 'DATE' | 'DATE_TIME';

export interface Schedule {
  date?: string | null;
  datePrecision?: DatePrecision | null;
  broadcasts?: string[] | null;
}

export function isEmptySchedule(schedule: Schedule): boolean {
  return !schedule.date && !schedule.datePrecision && (!schedule.broadcasts || schedule.broadcasts.length === 0);
}
